File: hackerrank/10 Days of Statistics/Day 0: Mean, Median, and Mode/solution.ex
defmodule Solution do
  n = IO.read(:stdio, :line)
  x = IO.read(:stdio, :line)

  n = n
  |> String.trim()
  |> String.to_integer()

  x = x
  |> String.split(" ")
  |> Enum.map(fn value -> String.to_integer(value) end)
  |> Enum.take(n)
  |> Enum.sort()

  mean = Enum.reduce(x, 0, fn value, acc -> value + acc end) / n
  middle = div(n, 2)

  median = if rem(n, 2) != 0 do
    Enum.at(x, middle)
  else
    (Enum.at(x, middle - 1) + Enum.at(x, middle)) / 2
  end

  [{mode, _} | _] = x
  |> Enum.reduce(
    %{},
  fn value, acc ->
    Map.update(acc, value, 1, fn c -> c + 1 end)
  end
  )
  |> Enum.sort(fn
    {x, a}, {y, b} when a == b -> x <= y
    {_, a}, {_, b} -> a >= b
  end)

  IO.puts(Float.round(mean, 1))
  IO.puts(Float.round(median, 1))
  IO.puts(mode)
end

File: hackerrank/10 Days of Statistics/Day 0: Mean, Median, and Mode/solution.c
#include <stdio.h>
#include <stdlib.h>

int cmpfunc(const void *a, const void *b) {
  return *(int*) a - *(int*) b;
}

void solution(const int n, const int *x) {
  int mode,
    most = 0,
    total = 0;
  double mean, median;

  qsort((void*) x, n, sizeof(int), cmpfunc);

  // mean

  for (int i = 0; i < n; i++) {
    total += x[i];
  }

  mean = total / (double) n;

  // median

  if (n % 2 == 0) {
    median = (x[n / 2 - 1] + x[n / 2]) / 2.0;
  } else {
    median = x[n / 2];
  }

  // mode

  for (int i = 0; i < n; i++) {
    int count = 1;

    while (x[i] == x[i + 1]) {
      i++;
      count++;
    }

    if (count > most) {
      mode = x[i];
      most = count;
    }
  }

  printf("%.6g\n%.6g\n%d\n", mean, median, mode);
}

int main() {
  int n, *x;

  scanf("%d", &n);
  x = (int*) malloc(n * sizeof(int));

  if (!x)
    return -1;

  for (int i = 0; i < n; i++) {
    scanf("%d", &x[i]);
  }

  solution(n, x);

  free(x);

  return 0;
}
